src/color.rs

```diff
@@ -562,109 +602,232 @@ fn parse_color_function<'i, 't, ComponentParser>(
 where
     ComponentParser: ColorComponentParser<'i>,
 {
-    let (red, green, blue, uses_commas) = match_ignore_ascii_case! { name,
+    let (red, green, blue, alpha) = match_ignore_ascii_case! { name,
         "rgb" | "rgba" => parse_rgb_components_rgb(component_parser, arguments)?,
-        "hsl" | "hsla" => parse_hsl_hwb(component_parser, arguments, hsl_to_rgb, /* allow_comma = */ true)?,
-        "hwb" => parse_hsl_hwb(component_parser, arguments, hwb_to_rgb, /* allow_comma = */ false)?,
+        "hsl" | "hsla" => parse_rgb_components_hsl(component_parser, arguments)?,
+        "hwb" => parse_rgb_components_hwb(component_parser, arguments)?,
         _ => return Err(arguments.new_unexpected_token_error(Token::Ident(name.to_owned().into()))),
     };
 
-    let alpha = if !arguments.is_exhausted() {
-        if uses_commas {
-            arguments.expect_comma()?;
-        } else {
-            arguments.expect_delim('/')?;
-        };
-        clamp_unit_f32(
-            component_parser
-                .parse_number_or_percentage(arguments)?
-                .unit_value(),
-        )
-    } else {
-        255
-    };
-
     arguments.expect_exhausted()?;
     Ok(rgba(red, green, blue, alpha))
 }
 
+#[inline]
+fn parse_alpha_component<'i, 't, ComponentParser>(
+    component_parser: &ComponentParser,
+    arguments: &mut Parser<'i, 't>,
+) -> Result<f32, ParseError<'i, ComponentParser::Error>>
+where
+    ComponentParser: ColorComponentParser<'i>,
+{
+    Ok(if !arguments.is_exhausted() {
+        arguments.expect_delim('/')?;
+
+        component_parser.parse_alpha_value(arguments)?.number
+    } else {
+        1.
+    })
+}
+
+/// https://w3c.github.io/csswg-drafts/css-color-4/#rgb-functions
 #[inline]
 fn parse_rgb_components_rgb<'i, 't, ComponentParser>(
     component_parser: &ComponentParser,
     arguments: &mut Parser<'i, 't>,
-) -> Result<(u8, u8, u8, bool), ParseError<'i, ComponentParser::Error>>
+) -> Result<(u8, u8, u8, u8), ParseError<'i, ComponentParser::Error>>
 where
     ComponentParser: ColorComponentParser<'i>,
 {
-    // Either integers or percentages, but all the same type.
-    // https://drafts.csswg.org/css-color/#rgb-functions
-    let (red, is_number) = match component_parser.parse_number_or_percentage(arguments)? {
-        NumberOrPercentage::Number { value } => (clamp_floor_256_f32(value), true),
-        NumberOrPercentage::Percentage { unit_value } => (clamp_unit_f32(unit_value), false),
-    };
+    // Try to parse legacy syntax first.
+    let legacy_syntax_result = arguments.try_parse(|input| {
```

Comment:
This function ends up being very long and duplicates quite a bit of code.  That and the fact that the hsl and hwb versions are also on their own now, duplicating code, I'm not sure this is an improvement over the original.

The abstractions for `Hue` and `Alpha` doesn't add much value on their own.  If you are the consumer of the parse_hue function you don't even know what unit you're getting back.  Adding some unit function might help, e.g. `Hue::as_degrees(..)`, `Hue::as_percentage(..)` or `Hue::as_unit(..)`.  This of course is that `NumberOrPercentage` does with enums.